File: services/py/quant_core/domain/rules/screen_eval.py
"""Evaluate a screening predicate against one stock's K-line slice.

Pure interpreter over a list of dict rows (one per trade_date, ordered
ascending). Adapters in the service layer materialise a Polars / Arrow
slice and pass the per-code rows here.

We keep this in the domain layer because it has zero IO and only
depends on :mod:`quant_core.domain.types.screen`. The simple
list-of-dict shape is intentional: it sidesteps Polars-version
sensitivity for the v1 cut. Performance is fine for the v1 budgets
(~5500 codes x ≤60 rows x ≤10 columns).
"""

from __future__ import annotations

from decimal import Decimal
from typing import TYPE_CHECKING

from quant_core.domain.types.screen import (
    Aggregate,
    Compare,
    Consecutive,
    Const,
    Exists,
    Field,
    ForAll,
    Logical,
    PeriodReturn,
)
from quant_core.errors import QuantError

if TYPE_CHECKING:
    from collections.abc import Mapping, Sequence

    from quant_core.domain.types.screen import Predicate, Scalar


# Sentinel for "no value at this row" (e.g. ma60 on a fresh listing).
_NA = object()


def evaluate_predicate(rows: Sequence[Mapping[str, object]], pred: Predicate) -> bool:
    """Return True iff ``pred`` matches given the stock's window slice.

    ``rows`` is sorted ascending by trade_date. ``rows[-1]`` is the
    ``asof`` bar; earlier rows are the lookback context.
    """
    if not rows:
        return False
    return _eval_predicate(rows, pred)


def _eval_predicate(rows: Sequence[Mapping[str, object]], pred: Predicate) -> bool:
    if isinstance(pred, Compare):
        return _eval_compare(rows, pred)
    if isinstance(pred, Logical):
        return _eval_logical(rows, pred)
    if isinstance(pred, ForAll):
        return _eval_for_all(rows, pred)
    if isinstance(pred, Exists):
        return _eval_exists(rows, pred)
    if isinstance(pred, Consecutive):
        return _eval_consecutive(rows, pred)
    raise QuantError(
        "EVALUATION_FAILED",
        f"unhandled predicate node: {type(pred).__name__}",
    )


def _eval_logical(rows: Sequence[Mapping[str, object]], node: Logical) -> bool:
    if node.op == "not":
        return not _eval_predicate(rows, node.args[0])
    if node.op == "and":
        return all(_eval_predicate(rows, a) for a in node.args)
    return any(_eval_predicate(rows, a) for a in node.args)


def _eval_compare(rows: Sequence[Mapping[str, object]], node: Compare) -> bool:
    left = _eval_scalar(rows, node.left)
    right = _eval_scalar(rows, node.right)
    if left is _NA or right is _NA:
        return False
    return _compare(node.op, left, right)


def _eval_for_all(rows: Sequence[Mapping[str, object]], node: ForAll) -> bool:
    if len(rows) < node.days:
        return False
    window = rows[-node.days :]
    return all(_eval_predicate(window[: i + 1], node.predicate) for i in range(len(window)))


def _eval_exists(rows: Sequence[Mapping[str, object]], node: Exists) -> bool:
    if len(rows) < node.days:
        return False
    window = rows[-node.days :]
    return any(_eval_predicate(window[: i + 1], node.predicate) for i in range(len(window)))


def _eval_consecutive(rows: Sequence[Mapping[str, object]], node: Consecutive) -> bool:
    streak = 0
    longest = 0
    for i in range(len(rows)):
        if _eval_predicate(rows[: i + 1], node.predicate):
            streak += 1
            longest = max(longest, streak)
        else:
            streak = 0
    return longest >= node.min_len


def _eval_scalar(rows: Sequence[Mapping[str, object]], node: Scalar) -> object:
    if isinstance(node, Field):
        return _row_value(rows[-1], node.field)
    if isinstance(node, Const):
        return node.value
    if isinstance(node, Aggregate):
        return _eval_aggregate(rows, node)
    if isinstance(node, PeriodReturn):
        return _eval_period_return(rows, node)
    raise QuantError("EVALUATION_FAILED", f"unhandled Scalar: {type(node).__name__}")


def _eval_aggregate(rows: Sequence[Mapping[str, object]], node: Aggregate) -> object:
    if len(rows) < node.days:
        return _NA
    window = rows[-node.days :]
    values: list[Decimal] = []
    for r in window:
        v = _row_value(r, node.field)
        if v is _NA:
            continue
        values.append(_to_decimal(v))
    if node.agg == "count":
        return Decimal(len(values))
    if not values:
        return _NA
    if node.agg == "mean":
        total = sum(values, Decimal(0))
        return total / len(values)
    if node.agg == "sum":
        return sum(values, Decimal(0))
    if node.agg == "min":
        return min(values)
    if node.agg == "max":
        return max(values)
    raise QuantError("EVALUATION_FAILED", f"unhandled agg: {node.agg!r}")


def _eval_period_return(rows: Sequence[Mapping[str, object]], node: PeriodReturn) -> object:
    if len(rows) < node.days + 1:
        return _NA
    end = _row_value(rows[-1], "close_qfq")
    start = _row_value(rows[-node.days - 1], "close_qfq")
    if end is _NA or start is _NA:
        return _NA
    start_dec = _to_decimal(start)
    if start_dec == 0:
        return _NA
    return (_to_decimal(end) - start_dec) / start_dec


def _row_value(row: Mapping[str, object], field: str) -> object:
    if field not in row:
        return _NA
    v = row[field]
    return _NA if v is None else v


def _compare(op: str, left: object, right: object) -> bool:
    a = _to_decimal(left)
    b = _to_decimal(right)
    if op == "gt":
        return a > b
    if op == "lt":
        return a < b
    if op == "gte":
        return a >= b
    if op == "lte":
        return a <= b
    if op == "eq":
        return a == b
    if op == "neq":
        return a != b
    raise QuantError("EVALUATION_FAILED", f"unhandled compare op: {op!r}")


def _to_decimal(v: object) -> Decimal:
    if isinstance(v, Decimal):
        return v
    if isinstance(v, bool):
        raise QuantError("EVALUATION_FAILED", "bool is not numeric")
    if isinstance(v, (int, float, str)):
        return Decimal(str(v))
    raise QuantError(
        "EVALUATION_FAILED",
        f"cannot coerce {type(v).__name__} to Decimal",
    )

File: apps/api/src/modules/screen/domain/pure/screen-eval.ts
/**
 * Pure interpreter for the screening predicate DSL.
 *
 * Port of `services/py/quant_core/domain/rules/screen_eval.py`. Operates
 * on a per-code `ScreenRow[]` slice sorted ascending by `trade_date`;
 * `rows[rows.length - 1]` is the `asof` bar.
 *
 * Null sentinel: a row's column being `null` (missing ma60 on a fresh
 * listing, missing `pct_chg_qfq` on the first bar of the slice) makes
 * every Compare that touches it evaluate to false — matches Python's
 * `_NA` semantics.
 */

import { QuantError, type DslPredicate, type DslScalar } from '@quant/shared';

import { D, type Dec } from '../../../../common/decimal.js';
import type { AggOp, CompareOp } from './screen-fields.js';

export interface ScreenRow {
  readonly trade_date: string; // ISO YYYY-MM-DD
  readonly open_qfq: number;
  readonly high_qfq: number;
  readonly low_qfq: number;
  readonly close_qfq: number;
  readonly volume: number;
  readonly amount: number;
  readonly turnover_rate: number;
  readonly ma5: number | null;
  readonly ma10: number | null;
  readonly ma20: number | null;
  readonly ma60: number | null;
  readonly pct_chg_qfq: number | null;
}

/** Sentinel — Py's `_NA`. */
const NA: unique symbol = Symbol('NA');
type Na = typeof NA;

/**
 * Evaluate a predicate against `rows`. Empty `rows` → false (matches Py).
 */
export function evaluatePredicate(rows: readonly ScreenRow[], pred: DslPredicate): boolean {
  if (rows.length === 0) return false;
  return evalPredicate(rows, pred);
}

/**
 * Evaluate a Scalar; returns `null` for the "no value" case so callers
 * (ranking, evidence) can detect it without importing the sentinel.
 */
export function evaluateScalar(rows: readonly ScreenRow[], scalar: DslScalar): Dec | null {
  if (rows.length === 0) return null;
  const v = evalScalar(rows, scalar);
  return v === NA ? null : v;
}

function evalPredicate(rows: readonly ScreenRow[], pred: DslPredicate): boolean {
  switch (pred.kind) {
    case 'compare':
      return evalCompare(rows, pred);
    case 'logical':
      return evalLogical(rows, pred);
    case 'for_all':
      return evalForAll(rows, pred);
    case 'exists':
      return evalExists(rows, pred);
    case 'consecutive':
      return evalConsecutive(rows, pred);
    default: {
      const exhaustive: never = pred;
      throw new QuantError(
        'EVALUATION_FAILED',
        `unhandled predicate kind: ${JSON.stringify(exhaustive)}`,
        {},
      );
    }
  }
}

function evalLogical(
  rows: readonly ScreenRow[],
  node: Extract<DslPredicate, { kind: 'logical' }>,
): boolean {
  if (node.op === 'not') {
    const first = node.args[0];
    if (first === undefined) {
      throw new QuantError('EVALUATION_FAILED', "logical 'not' requires an arg", {});
    }
    return !evalPredicate(rows, first);
  }
  if (node.op === 'and') {
    for (const a of node.args) {
      if (!evalPredicate(rows, a)) return false;
    }
    return true;
  }
  // 'or'
  for (const a of node.args) {
    if (evalPredicate(rows, a)) return true;
  }
  return false;
}

function evalCompare(
  rows: readonly ScreenRow[],
  node: Extract<DslPredicate, { kind: 'compare' }>,
): boolean {
  const left = evalScalar(rows, node.left);
  const right = evalScalar(rows, node.right);
  if (left === NA || right === NA) return false;
  if (!isCompareOp(node.op)) {
    throw new QuantError('EVALUATION_FAILED', `unhandled compare op: ${node.op}`, {});
  }
  switch (node.op) {
    case 'gt':
      return left.gt(right);
    case 'lt':
      return left.lt(right);
    case 'gte':
      return left.gte(right);
    case 'lte':
      return left.lte(right);
    case 'eq':
      return left.eq(right);
    case 'neq':
      return !left.eq(right);
  }
}

function evalForAll(
  rows: readonly ScreenRow[],
  node: Extract<DslPredicate, { kind: 'for_all' }>,
): boolean {
  const days = node.window.days;
  if (rows.length < days) return false;
  const window = rows.slice(rows.length - days);
  for (let i = 0; i < window.length; i++) {
    if (!evalPredicate(window.slice(0, i + 1), node.predicate)) return false;
  }
  return true;
}

function evalExists(
  rows: readonly ScreenRow[],
  node: Extract<DslPredicate, { kind: 'exists' }>,
): boolean {
  const days = node.window.days;
  if (rows.length < days) return false;
  const window = rows.slice(rows.length - days);
  for (let i = 0; i < window.length; i++) {
    if (evalPredicate(window.slice(0, i + 1), node.predicate)) return true;
  }
  return false;
}

function evalConsecutive(
  rows: readonly ScreenRow[],
  node: Extract<DslPredicate, { kind: 'consecutive' }>,
): boolean {
  let streak = 0;
  let longest = 0;
  for (let i = 0; i < rows.length; i++) {
    if (evalPredicate(rows.slice(0, i + 1), node.predicate)) {
      streak += 1;
      if (streak > longest) longest = streak;
    } else {
      streak = 0;
    }
  }
  return longest >= node.min_len;
}

function evalScalar(rows: readonly ScreenRow[], node: DslScalar): Dec | Na {
  switch (node.kind) {
    case 'field':
      return rowValue(rows[rows.length - 1]!, node.field);
    case 'const':
      return new D(node.value);
    case 'agg':
      return evalAggregate(rows, node);
    case 'period_return':
      return evalPeriodReturn(rows, node.window.days);
    case 'scale': {
      const inner = evalScalar(rows, node.inner);
      if (inner === NA) return NA;
      return inner.mul(new D(node.factor));
    }
    default: {
      const exhaustive: never = node;
      throw new QuantError(
        'EVALUATION_FAILED',
        `unhandled scalar kind: ${JSON.stringify(exhaustive)}`,
        {},
      );
    }
  }
}

function evalAggregate(
  rows: readonly ScreenRow[],
  node: Extract<DslScalar, { kind: 'agg' }>,
): Dec | Na {
  const days = node.window.days;
  if (rows.length < days) return NA;
  const window = rows.slice(rows.length - days);
  const values: Dec[] = [];
  for (const r of window) {
    const v = rowValue(r, node.field);
    if (v === NA) continue;
    values.push(v);
  }
  if (!isAggOp(node.agg)) {
    throw new QuantError('EVALUATION_FAILED', `unhandled agg: ${node.agg}`, {});
  }
  if (node.agg === 'count') return new D(values.length);
  if (values.length === 0) return NA;
  switch (node.agg) {
    case 'mean': {
      const sum = values.reduce<Dec>((acc, v) => acc.add(v), new D(0));
      return sum.div(values.length);
    }
    case 'sum':
      return values.reduce<Dec>((acc, v) => acc.add(v), new D(0));
    case 'min':
      return values.reduce<Dec>((acc, v) => (v.lt(acc) ? v : acc), values[0]!);
    case 'max':
      return values.reduce<Dec>((acc, v) => (v.gt(acc) ? v : acc), values[0]!);
  }
}

function evalPeriodReturn(rows: readonly ScreenRow[], days: number): Dec | Na {
  if (rows.length < days + 1) return NA;
  const end = rows[rows.length - 1]!.close_qfq;
  const startRow = rows[rows.length - 1 - days];
  if (startRow === undefined) return NA;
  if (end === null || startRow.close_qfq === null) return NA;
  const startDec = new D(startRow.close_qfq);
  if (startDec.eq(0)) return NA;
  return new D(end).sub(startDec).div(startDec);
}

function rowValue(row: ScreenRow, field: string): Dec | Na {
  const v = (row as unknown as Record<string, unknown>)[field];
  if (v === undefined || v === null) return NA;
  if (typeof v === 'number') return new D(v);
  if (typeof v === 'string') return new D(v);
  return NA;
}

function isCompareOp(op: string): op is CompareOp {
  return op === 'gt' || op === 'lt' || op === 'gte' || op === 'lte' || op === 'eq' || op === 'neq';
}

function isAggOp(op: string): op is AggOp {
  return op === 'mean' || op === 'sum' || op === 'min' || op === 'max' || op === 'count';
}
